src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -1276,24 +1276,33 @@ private static uint HandleEventSendComplete(State state, ref StreamEvent evt)
                     complete = true;
                 }
 
-                if (canceled)
+                if (sendCompleteEvent.Canceled != 0)
                 {
                     state.SendState = SendState.Aborted;
+
+                    //
+                    // There are multiple reasons the send could have been cancelled:
+                    //   - Connection was aborted (either by transport or peer) => error-code already provided on the connection-level event
+                    //   - Stream's receive side was aborted by peer => already handled by HandleEventPeerRecvAborted
+                    //     and we will not set the exception due to complete == false
+                    //   - Stream's send side was aborted locally => no connection-level abort code and we return QuicOperationAbortException
+                    //
+                    abortException = ThrowHelper.GetConnectionAbortedException(state.ConnectionState.AbortErrorCode);
                 }
             }
 
             if (complete)
             {
                 CleanupSendState(state);
 
-                if (!canceled)
+                if (abortException == null)
                 {
                     state.SendResettableCompletionSource.Complete(MsQuicStatusCodes.Success);
                 }
                 else
                 {
                     state.SendResettableCompletionSource.CompleteException(
-                        ExceptionDispatchInfo.SetCurrentStackTrace(new OperationCanceledException("Write was canceled")));
+                        ExceptionDispatchInfo.SetCurrentStackTrace(abortException));
```

Comment:
Is there a reason to create the exception within the lock?
Could this whole change be reduced to just:
```suggestion
                        ExceptionDispatchInfo.SetCurrentStackTrace(ThrowHelper.GetConnectionAbortedException(state.ConnectionState.AbortErrorCode)));
```

Reply:
I originally had code dealing with stream write aborts as well, until I realized that they are handled in `HandleEventPeerRecvAborted` and will never get hit here. So if accessing `state.ConnectionState.AbortErrorCode` is fine outside of the lock, then I can simplify this.